File: transpiler/expected/minilexer.zig
const std = @import("std");
const mem = std.mem;
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;

var gpa = std.heap.GeneralPurposeAllocator(.{}){};
const allocator = gpa.allocator();

pub const EOF: i64 = -1;

pub const Token = struct {
    kind: []const u8,
    text: []const u8,
    pos: i64,

    pub fn is_word(self: *const Token) bool {
        return mem.eql(u8, self.kind, "word");
    }
};

pub const Lexer = struct {
    source: []const u8,
    pos: i64,
    current: ?Token,

    pub fn peek(self: *Lexer) i64 {
        if (@as(usize, @intCast(self.pos)) >= self.source.len) {
            return EOF;
        }
        return self.source[@as(usize, @intCast(self.pos))];
    }

    pub fn advance(self: *Lexer) void {
        self.pos += 1;
    }

    pub fn scan_word(self: *Lexer) struct { f0: Token, f1: bool } {
        const start = self.pos;
        while (self.peek() != EOF and !is_space(self.peek())) {
            self.advance();
        }
        if (self.pos == start) {
            return .{ .f0 = Token{}, .f1 = false };
        }
        const text: []const u8 = self.source[@as(usize, @intCast(start))..@as(usize, @intCast(self.pos))];
        return .{ .f0 = Token{ .kind = "word", .text = text, .pos = start }, .f1 = true };
    }
};

pub fn is_space(ch: i64) bool {
    return ch == ' ' or ch == '\n';
}

pub fn tokenize(source: []const u8) ArrayList(Token) {
    var lx = Lexer{ .source = source, .pos = 0, .current = null };
    var tokens: ArrayList(Token) = ArrayList(Token).init(allocator);
    while (lx.peek() != EOF) {
        const ch = lx.peek();
        if (is_space(ch)) {
            lx.advance();
            continue;
        }
        const result = lx.scan_word();
        const tok = result.f0;
        const ok = result.f1;
        if (!ok) {
            @panic("unexpected character");
        }
        tokens.append(tok) catch unreachable;
    }
    return tokens;
}

pub fn count_words(tokens: ArrayList(Token)) i64 {
    var count: i64 = 0;
    for (tokens.items) |tok| {
        if (mem.eql(u8, tok.kind, "word")) {
            count += 1;
        }
    }
    return count;
}

pub fn format_token(tok: Token) []const u8 {
    return tok.kind ++ ":" ++ tok.text;
}

pub fn find_token(tokens: ArrayList(Token), kind: []const u8) ?Token {
    for (tokens.items) |tok| {
        if (mem.eql(u8, tok.kind, kind)) {
            return tok;
        }
    }
    return null;
}

pub fn example_nil_check(tokens: ArrayList(Token)) []const u8 {
    const tok = find_token(tokens, "word");
    if (tok == null) {
        return "";
    }
    return tok.?.text;
}

pub fn sum_positions(tokens: ArrayList(Token)) i64 {
    var sum: i64 = 0;
    var i: i64 = 0;
    while (@as(usize, @intCast(i)) < tokens.items.len) {
        sum += tokens.items[@as(usize, @intCast(i))].pos;
        i += 1;
    }
    return sum;
}

pub fn first_word_pos(tokens: ArrayList(Token)) i64 {
    var pos: i64 = -1;
    for (tokens.items) |tok| {
        if (mem.eql(u8, tok.kind, "word")) {
            pos = tok.pos;
            break;
        }
    }
    return pos;
}

pub fn max_int(a: i64, b: i64) i64 {
    return if (a > b) a else b;
}

pub fn default_kinds() std.StringHashMap(i64) {
    return blk: { var m = std.StringHashMap(i64).init(allocator); m.put("word", 1) catch unreachable; m.put("num", 2) catch unreachable; m.put("op", 3) catch unreachable; break :blk m; };
}

pub fn scoped_work(x: i64) i64 {
    var result: i64 = 0;
    {
        const temp: i64 = x * 2;
        result = temp + 1;
    }
    return result;
}

pub fn kind_priority(kind: []const u8) i64 {
    if (mem.eql(u8, kind, "word")) {
        return 1;
    } else if (mem.eql(u8, kind, "num") or mem.eql(u8, kind, "float")) {
        return 2;
    } else if (mem.eql(u8, kind, "op")) {
        return 3;
    } else {
        return 0;
    }
}

pub fn safe_tokenize(source: []const u8) ArrayList(Token) {
    var tokens: ArrayList(Token) = ArrayList(Token).init(allocator);
    // Note: error handling simplified - Zig uses error unions
    tokens = tokenize(source);
    return tokens;
}

pub fn pi() f64 {
    return 3.14159;
}

pub fn describe_token(tok: Token) []const u8 {
    return std.fmt.allocPrint(allocator, "Token({}, {}, {})", .{ tok.kind, tok.text, tok.pos }) catch unreachable;
}

pub fn set_first_kind(tokens: ArrayList(Token), kind: []const u8) void {
    if (tokens.items.len > 0) {
        tokens.items[0] = Token{ .kind = kind, .text = "", .pos = 0 };
    }
}

pub fn make_int_slice(n: i64) ArrayList(i64) {
    return ArrayList(i64).initCapacity(allocator, @as(usize, @intCast(n))) catch unreachable;
}

pub fn int_to_float(n: i64) f64 {
    return @as(f64, @floatFromInt(n));
}

pub fn known_kinds() std.StringHashMap(void) {
    return blk: { var s = std.StringHashMap(void).init(allocator); s.put("word", {}) catch unreachable; s.put("num", {}) catch unreachable; s.put("op", {}) catch unreachable; break :blk s; };
}

pub fn new_kind_map() std.StringHashMap(i64) {
    return std.StringHashMap(i64).init(allocator);
}

pub fn get_array_first(arr: [10]i64) i64 {
    return arr[0];
}

pub fn maybe_get(tokens: ArrayList(Token), idx: i64) ?Token {
    if (@as(usize, @intCast(idx)) >= tokens.items.len) {
        return null;
    }
    return tokens.items[@as(usize, @intCast(idx))];
}

pub fn set_via_ptr(ptr: *i64, val: i64) void {
    ptr.* = val;
}

pub fn identity_str(s: []const u8) []const u8 {
    return s;
}

pub fn accept_union(_: union(enum) { v0: Token, v1: Lexer }) bool {
    return true;
}

File: transpiler/expected/minilexer.go
package minilexer

import (
	"fmt"
	"reflect"
	"strconv"
	"strings"
	"unicode"
	"unicode/utf8"
)

var _ = strings.Compare // ensure import is used

func _strIsAlnum(s string) bool {
	for _, r := range s {
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) {
			return false
		}
	}
	return len(s) > 0
}

func _strIsAlpha(s string) bool {
	for _, r := range s {
		if !unicode.IsLetter(r) {
			return false
		}
	}
	return len(s) > 0
}

func _strIsDigit(s string) bool {
	for _, r := range s {
		if !unicode.IsDigit(r) {
			return false
		}
	}
	return len(s) > 0
}

func _strIsSpace(s string) bool {
	for _, r := range s {
		if !unicode.IsSpace(r) {
			return false
		}
	}
	return len(s) > 0
}

func _strIsUpper(s string) bool {
	for _, r := range s {
		if !unicode.IsUpper(r) {
			return false
		}
	}
	return len(s) > 0
}

func _strIsLower(s string) bool {
	for _, r := range s {
		if !unicode.IsLower(r) {
			return false
		}
	}
	return len(s) > 0
}

// _intPtr converts a sentinel int (-1 = nil) to *int
func _intPtr(val int) *int {
	if val == -1 {
		return nil
	}
	return &val
}

// Range generates a slice of integers similar to Python's range()
// Range(end) -> [0, 1, ..., end-1]
// Range(start, end) -> [start, start+1, ..., end-1]
// Range(start, end, step) -> [start, start+step, ..., last < end]
func Range(args ...int) []int {
	var start, end, step int
	switch len(args) {
	case 1:
		start, end, step = 0, args[0], 1
	case 2:
		start, end, step = args[0], args[1], 1
	case 3:
		start, end, step = args[0], args[1], args[2]
	default:
		return nil
	}
	if step == 0 {
		return nil
	}
	var result []int
	if step > 0 {
		for i := start; i < end; i += step {
			result = append(result, i)
		}
	} else {
		for i := start; i > end; i += step {
			result = append(result, i)
		}
	}
	return result
}

func _parseInt(s string, base int) int {
	n, _ := strconv.ParseInt(s, base, 64)
	return int(n)
}

// _mapGet returns the value for key from map m, or defaultVal if not found
func _mapGet[K comparable, V any](m map[K]V, key K, defaultVal V) V {
	if v, ok := m[key]; ok {
		return v
	}
	return defaultVal
}

// _intToStr converts an integer to its string representation
func _intToStr(n int) string {
	return strconv.Itoa(n)
}

// _isNilInterface checks if an interface value is nil.
// In Go, an interface is nil only when both type and value are nil.
// This handles the case where interface contains a typed nil pointer.
func _isNilInterface(i interface{}) bool {
	if i == nil {
		return true
	}
	v := reflect.ValueOf(i)
	return v.Kind() == reflect.Ptr && v.IsNil()
}

// _runeAt returns the character (as string) at rune index i in string s.
// Python s[i] on a string returns the i-th character, not byte.
func _runeAt(s string, i int) string {
	runes := []rune(s)
	if i < 0 || i >= len(runes) {
		return ""
	}
	return string(runes[i])
}

// _runeLen returns the number of runes (characters) in string s.
// Python len(s) on a string returns character count, not byte count.
func _runeLen(s string) int {
	return utf8.RuneCountInString(s)
}

// _Substring returns s[start:end] using rune (character) indices.
// Python s[start:end] uses character indices, not byte indices.
func _Substring(s string, start int, end int) string {
	runes := []rune(s)
	if start < 0 {
		start = 0
	}
	if end > len(runes) {
		end = len(runes)
	}
	if start >= end {
		return ""
	}
	return string(runes[start:end])
}

const (
	EOF = -1
)

type Scanner interface {
	peek() int
	advance()
}

type Token struct {
	Kind string
	Text string
	Pos int
}

func (t *Token) IsWord() bool {
	return t.Kind == "word"
}

type Lexer struct {
	Source string
	Pos int
	Current *Token
}

func (lx *Lexer) Peek() int {
	if lx.Pos >= _runeLen(lx.Source) {
		return EOF
	}
	return lx.Source[lx.Pos]
}

func (lx *Lexer) Advance()  {
	lx.Pos++
}

func (lx *Lexer) ScanWord() (Token, bool) {
	start := lx.Pos
	for lx.Peek() != EOF && !IsSpace(lx.Peek()) {
		lx.Advance()
	}
	if lx.Pos == start {
		return Token{}, false
	}
	text := _Substring(lx.Source, start, lx.Pos)
	return Token{Kind: "word", Text: text, Pos: start}, true
}

func IsSpace(ch int) bool {
	return ch == 32 || ch == 10
}

func Tokenize(source string) []Token {
	lx := Lexer{Source: source, Pos: 0, Current: nil}
	tokens := []Token{}
	for lx.Peek() != EOF {
		ch := lx.Peek()
		if IsSpace(ch) {
			lx.Advance()
			continue
		}
		var result struct{ F0 Token; F1 bool } = func() struct{ F0 Token; F1 bool } { _t0, _t1 := lx.ScanWord(); return struct{ F0 Token; F1 bool }{_t0, _t1} }()
		tok := result.F0
		ok := result.F1
		if !ok {
			panic(NewParseError("unexpected character", lx.Pos, 0))
		}
		tokens = append(tokens, tok)
	}
	return tokens
}

func CountWords(tokens []Token) int {
	count := 0
	for _, tok := range tokens {
		if tok.Kind == "word" {
			count++
		}
	}
	return count
}

func FormatToken(tok Token) string {
	return tok.Kind + ":" + tok.Text
}

func FindToken(tokens []Token, kind string) *Token {
	for _, tok := range tokens {
		if tok.Kind == kind {
			return &tok
		}
	}
	return nil
}

func ExampleNilCheck(tokens []Token) string {
	tok := FindToken(tokens, "word")
	if tok == nil {
		return ""
	}
	return tok.Text
}

func SumPositions(tokens []Token) int {
	sum := 0
	for i := 0; i < len(tokens); i++ {
		sum += tokens[i].Pos
	}
	return sum
}

func FirstWordPos(tokens []Token) int {
	pos := -1
	for _, tok := range tokens {
		if tok.Kind == "word" {
			pos = tok.Pos
			break
		}
	}
	return pos
}

func MaxInt(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func DefaultKinds() map[string]int {
	return map[string]int{"word": 1, "num": 2, "op": 3}
}

func ScopedWork(x int) int {
	result := 0
	{
		temp := x * 2
		result = temp + 1
	}
	return result
}

func KindPriority(kind string) int {
	switch kind {
	case "word":
		return 1
	case "num", "float":
		return 2
	case "op":
		return 3
	default:
		return 0
	}
}

func SafeTokenize(source string) []Token {
	tokens := []Token{}
	func() {
		defer func() {
			if e := recover(); e != nil {
				tokens = []Token{}
			}
		}()
		tokens = Tokenize(source)
	}()
	return tokens
}

func Pi() float64 {
	return 3.14159
}

func DescribeToken(tok Token) string {
	return fmt.Sprintf("Token(%v, %v, %v)", tok.Kind, tok.Text, tok.Pos)
}

func SetFirstKind(tokens []Token, kind string) {
	if len(tokens) > 0 {
		tokens[0] = Token{Kind: kind, Text: "", Pos: 0}
	}
}

func MakeIntSlice(n int) []int {
	return make([]int, n)
}

func IntToFloat(n int) float64 {
	return float64(n)
}

func KnownKinds() map[string]struct{} {
	return map[string]struct{}{"word": {}, "num": {}, "op": {}}
}

func NewKindMap() map[string]int {
	return make(map[string]int)
}

func GetArrayFirst(arr [10]int) int {
	return arr[0]
}

func MaybeGet(tokens []Token, idx int) *Token {
	if idx >= len(tokens) {
		return nil
	}
	return &tokens[idx]
}

func SetViaPtr(ptr *int, val int) {
	*ptr = val
}

func IdentityStr(s string) string {
	return s
}

func AcceptUnion(obj interface{}) bool {
	return true
}

